kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs

```diff
@@ -921,17 +915,10 @@ impl<'tcx> GotocCtx<'tcx> {
         let cast_src_ptr = src_ptr.clone().cast_to(Type::ssize_t());
         let offset = cast_dst_ptr.sub(cast_src_ptr);
 
-        // Check that computing `offset_bytes` would not overflow an `isize`
+        // Check that computing `offset` in bytes would not overflow an `isize`
         let ty = self.monomorphize(instance.substs.type_at(0));
-        let layout = self.layout_of(ty);
-        let size = Expr::int_constant(layout.size.bytes(), Type::ssize_t());
-        let offset_bytes = offset.clone().mul_overflow(size);
-        let overflow_check = self.codegen_assert(
-            offset_bytes.overflowed.not(),
-            PropertyClass::ArithmeticOverflow,
-            "attempt to compute offset in bytes which would overflow",
-            loc,
-        );
+        let (_offset_bytes, overflow_check) =
+            self.count_in_bytes(offset.clone(), ty, Type::ssize_t(), "ptr_offset_from", loc);
```

Comment:
If I understood the documentation correctly (https://doc.rust-lang.org/std/primitive.pointer.html#method.offset_from), I think we need to divide (and not multiply) the offset between the two pointers (`offset`) by the size of `ty`.

Reply:
I think there is a typo there and the documentation should say:
> The returned value is in units of T: the distance in bytes divided by `mem::size_of::<T>()`

The offset between two pointers is usually given in units of `T` (compare with the documentation for [`offset`](https://doc.rust-lang.org/std/primitive.pointer.html#method.offset)), which you can convert to bytes by multiplying
`offset * mem::size_of<T>` (what we do here).

I've opened [this PR](https://github.com/rust-lang/rust/pull/96639) to correct the typo.

Reply:
Isn't the offset computes using:
```
  let offset = cast_dst_ptr.sub(cast_src_ptr);
```
in terms of bytes? If so, then to convert it to be in terms of units of `T`, we need to divide by size of `T` since "offset in terms of bytes = offset in terms of T x size of T".

Reply:
No, offset computation is done in terms of T. See the test case in `tests/kani/PointerOffset/offset_from.rs`:

```rust
    let a = [0; 5];
    let b = [0; 5];
    let ptr1: *const i32 = &a[1];
    let ptr2: *const i32 = &a[3];
    let ptr3: *const i64 = &b[1];
    let ptr4: *const i64 = &b[3];
    unsafe {
        assert!(ptr2.offset_from(ptr1) == 2);
        assert!(ptr1.offset_from(ptr2) == -2);
        assert!(ptr4.offset_from(ptr3) == 2);
        assert!(ptr3.offset_from(ptr4) == -2);
    }
```
Here, assertion 1 and 3 (or 2 and 4) return the same offset value. An offset computed in terms of bytes would return 8 in the first one (2 * 4 bytes for each `i32` value) and 16 on assertion 3 (2 * 8 bytes for each `i64` value).

Reply:
One thing I was missing is that the return value is computed separately on line 937/924, and is not computed from `count_in_bytes`. I'm not sure I understand why an overflow check is needed here though. I'm assuming that a pointer subtraction in CBMC would involve a division operation and not a multiplication, i.e. `x-y` translates to `(x-y)/sizeof(*x)`?

Reply:
The overflow check is done because it's one of the safety conditions listed in the documentation:

> The distance between the pointers, in bytes, cannot overflow an isize.

I don't know for sure how CBMC handles pointer arithmetic internally.

Reply:
I see. But is this what the code is doing? I guess here's my source of confusion: what does `offset` represent in the lines below?
```
        let dst_ptr = fargs.remove(0);
        let src_ptr = fargs.remove(0);
        // Compute the offset with standard substraction using `isize`
        let cast_dst_ptr = dst_ptr.clone().cast_to(Type::ssize_t());
        let cast_src_ptr = src_ptr.clone().cast_to(Type::ssize_t());
        let offset = cast_dst_ptr.sub(cast_src_ptr);
```

Reply:
@zhassan-aws pointed out in a discussion that, because of these conversions, the computed offset would be in bytes and not units of T. So the overflow check should be based on this `sub` operation, and using `count_in_bytes` does not proceed here.

Thanks @zhassan-aws !